src/string/aho_corasick_test.cpp

```diff
@@ -0,0 +1,44 @@
+#include <gtest/gtest.h>
+#include "aho_corasick.cpp"
+
+// TODO: add more tests
+
+string gen(int n, int k) {
+	string s;
+	rep(i, 0, n)
+		s.push_back('a' + rand() % k);
+	return s;
+}
+
+set<pair<int, int>> se;
+
+void occurence(int i, int index) {
+	EXPECT_FALSE(se.count({i, index}));
+	se.insert({i, index});
+}
+
+bool matches(string hay, string needle, int start) {
```

Comment:
strings here could be references
also - bound checking for `hay`? Maybe something like `return needle == hay.substr(start, needle.size())` would be simpler (+some bound checking)

Reply:
Oh, thanks, forgot about bound checking :blush:. I'll rewrite it.